src/cplscheme/BiCouplingScheme.cpp

```diff
@@ -121,5 +127,51 @@ CouplingData *BiCouplingScheme::getReceiveData(
   return nullptr;
 }
 
+void BiCouplingScheme::exchangeInitialData()
+{
+  // F: send, receive, S: receive, send
+  if (doesFirstStep()) {
+    if (sendsInitializedData()) {
+      sendData(getM2N(), getSendData());
+    }
+    if (receivesInitializedData()) {
+      receiveData(getM2N(), getReceiveData());
+      checkInitialDataHasBeenReceived();
+    }
+  } else { // second participant
+    if (receivesInitializedData()) {
+      receiveData(getM2N(), getReceiveData());
+      checkInitialDataHasBeenReceived();
+    }
+    if (sendsInitializedData()) {
+      sendData(getM2N(), getSendData());
+    }
+  }
+}
+
+bool BiCouplingScheme::exchangeDataAndAccelerateImpl()
+{
+  bool convergence = true;
+  if (doesFirstStep()) { // first participant
+    PRECICE_DEBUG("Sending data...");
+    sendData(getM2N(), getSendData());
+    PRECICE_DEBUG("Receiving data...");
+    if (isImplicitCouplingScheme()) {
+      convergence = receiveConvergence(getM2N());
+    }
+    receiveData(getM2N(), getReceiveData());
+    checkDataHasBeenReceived();
+  } else { // second participant
+    if (isImplicitCouplingScheme()) {
+      PRECICE_DEBUG("Perform acceleration (only second participant)...");
+      convergence = doImplicitStep();
+      sendConvergence(getM2N(), convergence);
+    }
+    PRECICE_DEBUG("Sending data...");
+    sendData(getM2N(), getSendData());
+  }
```

Comment:
Where did the `receiveData` go here? Let's not hide this in `doImplicitStep()` (if that's the case)
How can this function be the same for both serial and parallel coupling?

Reply:
I reverted this refactoring. Should be clearer now.